### PR TITLE
[CPU][ArmSME] Enable transposes for f32 and f64

### PR DESCRIPTION
**Pull request by c-rhodes**
This patch enables transposes for ArmSME for f32 and f64 types now that they
may get eliminated by folding into defining `vector.transfer_read` since [1].
This is done by extending `setTransposeLikeOpRootConfig` which currently only
supports x86 (no changes there). The transpose is represented by a
`linalg.generic`, since `linalg.transpose` gets converted by
`GeneralizeLinalgNamedOps`.

[1] https://github.com/llvm/llvm-project/pull/92562

ci-extra: build_test_all_arm64

### CODE REVIEW

**Comment by MacDue** [comment]
Can you add `ci-extra: build_test_all_arm64` as the last line of your PR summary (and push something to re-trigger CI)?

**Review by MaheshRavishankar — COMMENTED** [review]
Would be good to wait for either @hanhanW or @dcaballe review.

**Comment by c-rhodes** [comment]
> Can you add `ci-extra: build_test_all_arm64` as the last line of your PR summary (and push something to re-trigger CI)?

Looks like CI ran but I'll add it to summary for any updates.

**Comment by c-rhodes** [comment]
> > Can you add `ci-extra: build_test_all_arm64` as the last line of your PR summary (and push something to re-trigger CI)?
> 
> Looks like CI ran but I'll add it to summary for any updates.

Just noticed `ci-extra: build_test_all_arm64` was skipped, will push to trigger.

**Comment by c-rhodes** [comment]
I've moved this to draft, will re-open once https://github.com/llvm/llvm-project/pull/92562 lands in the next LLVM integration.

File: compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp
Comment (MacDue) [inline compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp]: I wonder if this should be more restrictive? This seems to set 2-D scalable tile sizes for any transpose-like operation, but not all of them can/do lower to ArmSME, for example, I tired out:
```
  %empty = tensor.empty(): tensor<2x16x16xf32>
  %rhs = linalg.transpose ins(%transposed_b : tensor<2x16x16xf32>)
      outs(%empty : tensor<2x16x16xf32>) permutation = [0, 2, 1]
```
...it set 2-D scalable flags but ended up unvectorized (and used linalg-to-loops in the end).

Comment (c-rhodes) [reply compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp]: absolutely, I misinterpreted `x86TransposeLoweringPrecondition` to be restricted to 2 dims. I've kept that as is and created a separate one for SME to be more restrictive, test also added.

File: compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp
Comment (hanhanW) [inline compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp]: The big if-else makes readability really bad. Can we create methods to get `minTileSizes`? So it can be like `std::optional<SmallVector<int64_t>> getX86TransposeVectorSizes()` and `std::optional<SmallVector<int64_t>> getAArch64TransposeVectorSizes()`. What do you think?

File: compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp
Comment (MacDue) [inline compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp]: I think this also needs to check the body of the generic is simply a `scf.yield %in`. 
I tried: 
```
linalg.generic {indexing_maps = [affine_map<(d0, d1) -> (d1, d0)>, affine_map<(d0, d1) -> (d0, d1)>], iterator_types = ["parallel", "parallel"]} ins(%2 : tensor<32x32xf32>) outs(%3 : tensor<32x32xf32>) {
^bb0(%in: f32, %out: f32):
  %mul = arith.mulf %in, %in : f32
  linalg.yield %mul : f32
} -> tensor<32x32xf32>
```

And that's still a 'transpose'  and it adds the scalable flags and you get: 
```
%11 = arith.mulf %9, %9 : vector<[4]x[4]xf32>
%12 = vector.transpose %11, [1, 0] : vector<[4]x[4]xf32> to vector<[4]x[4]xf32>
%13 = vector.transfer_write %12, %extracted_slice_0[%c0, %c0], %10 {in_bounds = [true, true]} : vector<[4]x[4]xf32>, tensor<?x?xf32>
```

Comment (MacDue) [reply compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp]: (`opKnownToSupport2DScalableVectorizationWithArmSME` would also need to check this)

**Review by banach-space — COMMENTED** [review]
Thanks Cullen!

Sorry for the delay, I've kept skipping this one without realising that this is _the patch_ :) I blame the title:
> [CPU][ArmSME] Enable transposes for f32 and f64

That's not what's happening here, is it? IIUC, this patch adds logic to identify transpose-like `linalg.generic` Ops and then, for ArmSME, adds "scalable" flags to the tiling config.  For X86 there should be no changes, right? 

That's my understanding based on the code - is this correct? If yes, could you update the summary?

File: compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp
Comment (c-rhodes) [reply compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp]: good spot! I've updated to check for this, but had to move this function to utils to also call it from `opKnownToSupport2DScalableVectorizationWithArmSME`.

**Comment by c-rhodes** [comment]
> Thanks Cullen!
> 
> Sorry for the delay, I've kept skipping this one without realising that this is _the patch_ :) I blame the title:
> 
> > [CPU][ArmSME] Enable transposes for f32 and f64
> 
> That's not what's happening here, is it? IIUC, this patch adds logic to identify transpose-like `linalg.generic` Ops and then, for ArmSME, adds "scalable" flags to the tiling config. For X86 there should be no changes, right?
> 
> That's my understanding based on the code - is this correct? If yes, could you update the summary?



> Thanks Cullen!
> 
> Sorry for the delay, I've kept skipping this one without realising that this is _the patch_ :) I blame the title:
> 
> > [CPU][ArmSME] Enable transposes for f32 and f64
> 
> That's not what's happening here, is it? IIUC, this patch adds logic to identify transpose-like `linalg.generic` Ops and then, for ArmSME, adds "scalable" flags to the tiling config. For X86 there should be no changes, right?

It is what's happening, albeit indirectly since `linalg.transpose` ops are converted to `linalg.generic` by GeneralizeLinalgNamedOps`. But yes, no changes to X86.

> That's my understanding based on the code - is this correct? If yes, could you update the summary?

Not quite, but I'll clarify in the summary to make it clearer 👍

File: compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp
Comment (banach-space) [inline compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp]: I think that you also need to make sure that for element types other than `f32` and `f64`, the pre-condition fails. In the spirit of failing early.

**Comment by banach-space** [comment]
> Would be good to wait for either @hanhanW or @dcaballe review.

AFAIK, Diego is distracted with other activities these days :)

This looks good to me, what do you think @hanhanW and @MacDue ?

**Review by banach-space — APPROVED** [review]
LGTM, modulo handling other data types, thanks!

File: compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp
Comment (MacDue) [reply compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp]: _technically_ this fails no later than if it were in the pre-condition (but avoids duplicating it there, so I think this is fine).

**Review by MacDue — APPROVED** [review]
LGTM :+1: Thanks for the changes!

File: compiler/src/iree/compiler/Codegen/LLVMCPU/Utils.cpp
Comment (MacDue) [inline compiler/src/iree/compiler/Codegen/LLVMCPU/Utils.cpp]: I think calling this `transposeLoweringPreconditionAArch64SME` makes this look more specific to ArmSME than it really is. This is just something like `isLinalgGeneric2DTransposition`.

Comment (hanhanW) [reply compiler/src/iree/compiler/Codegen/LLVMCPU/Utils.cpp]: +1, I don't see any ArmSME specifics in the method.

File: compiler/src/iree/compiler/Codegen/LLVMCPU/Utils.cpp
Comment (hanhanW) [inline compiler/src/iree/compiler/Codegen/LLVMCPU/Utils.cpp]: nit: I think we can use `cast` because it should already be verified by LinalgOp.

File: compiler/src/iree/compiler/Codegen/LLVMCPU/Utils.cpp
Comment (hanhanW) [inline compiler/src/iree/compiler/Codegen/LLVMCPU/Utils.cpp]: nit: we don't need `()` for the two conditions.

```suggestion
  if (genericOp.getNumDpsInputs() != 1 || genericOp.getNumDpsInits() != 1)
    return false;
```

File: compiler/src/iree/compiler/Codegen/LLVMCPU/Utils.cpp
Comment (hanhanW) [inline compiler/src/iree/compiler/Codegen/LLVMCPU/Utils.cpp:66]: nit: remove the comments, because the code already describes what they are doing. Same comment for line 70.

https://google.github.io/styleguide/cppguide#Implementation_Comment_Donts

File: compiler/src/iree/compiler/Codegen/LLVMCPU/Utils.cpp
Comment (hanhanW) [inline compiler/src/iree/compiler/Codegen/LLVMCPU/Utils.cpp]: Given that we already check that the number of inits is one, so we don't need the for loop here. (I believe that people won't expect that there are more than one value, so it's fine.)

File: compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp
Comment (hanhanW) [inline compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp]: This kind of logics keep showing up in targets that don't care about scalable vectors. I did not get time to revamp the idea/design of large refactoring with @banach-space that I mentioned in https://github.com/iree-org/iree/pull/16435/files#r1513504805 (my bad). We really need a plan and take action items to address the issue. What do you think about it? (I'm just flagging the concern, but not saying that it is a blocker.)

File: compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp
Comment (hanhanW) [inline compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp]: nit: use `.isF32`. same for `.isF64`.

File: compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp
Comment (hanhanW) [inline compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp]: Need more details about the documentation. Without looking into implementation, it is not clear that the vectors are empty when it fails. Also, how about returning a boolean value which indicates if it succeeds or not?

File: compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp
Comment (hanhanW) [inline compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp]: few nits:

1. It is better to use `isX86` for the first condition.
2. style nit: wrap `braces` for the if-else statements. This is the few difference between IREE and MLIR in terms of coding style.
3. Can we move the `isSMEFeature` and `clEnableScalableVectorization` logics to `getTransposeAArch64SMEVectorSizes`, and rename it to `getTransposeAArch64VectorSizes`? Today, most of the methods are separated based on the `isX86`, `isRISCV`, and `isAArch64`. IMO, keeping it this way is better for layering. I'm afraid that this if-statements becomes a monster and adds complexities.

**Review by hanhanW — CHANGES_REQUESTED** [review]

File: compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp
Comment (c-rhodes) [reply compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp]: I quite like the empty state of `sizes` indicating failure so I've kept that, but updated the docs to make it clear 👍

File: compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp
Comment (c-rhodes) [reply compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp]: strongly agree targets that don't support scalability shouldn't have to care about it, sloppiness on my part. I've updated the target agnostic `getTransposeVectorSizes` to assume empty scalable flags indicates the target doesn't care, so target-specific hooks like this don't have to deal with it.

File: compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp
Comment (c-rhodes) [reply compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp]: > nit: use `.isF32`. same for `.isF64`.

doesn't compile

File: compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp
Comment (c-rhodes) [reply compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp]: much nicer 👍

File: compiler/src/iree/compiler/Codegen/LLVMCPU/Utils.cpp
Comment (c-rhodes) [reply compiler/src/iree/compiler/Codegen/LLVMCPU/Utils.cpp:66]: personally find they're much quicker to parse, so I've them.

File: compiler/src/iree/compiler/Codegen/LLVMCPU/LLVMCPU2DScalableTo1DScalable.cpp
Comment (MacDue) [inline compiler/src/iree/compiler/Codegen/LLVMCPU/LLVMCPU2DScalableTo1DScalable.cpp]: `LinalgOpInfo` is redundant right? 
```suggestion
    return isLinalgGeneric2DTranspose(genericOp);
```

File: compiler/src/iree/compiler/Codegen/LLVMCPU/Utils.cpp
Comment (banach-space) [inline compiler/src/iree/compiler/Codegen/LLVMCPU/Utils.cpp]:
```suggestion
  // The yield op should return the block argument corresponding to the input.
```

File: compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp
Comment (banach-space) [inline compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp:2015]: This _will_ return empty sizes for SVE - what's the plan for SVE here? Perhaps just add a TODO?

File: compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp
Comment (banach-space) [reply compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp]: > We really need a plan and take action items to address the issue. What do you think about it?

@hanhanW 100% agree, sorry for not being more pro-active about this :( I was waiting for a bit more feedback, but I think that the overall suggestion was to try a bit more substantial refactor?

File: compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp
Comment (c-rhodes) [reply compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp:2015]: I've added a note that only SME is supported. The focus of this patch is on SME, this could be extended to support SVE.

File: compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp
Comment (hanhanW) [reply compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp]: oh I see, it is because that `nonWideningLinalgElementType` returns `FailureOr<Type>` but not `Type`. I misread it, sorry for the bad suggestion.

File: compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp
Comment (hanhanW) [inline compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp]: I think the type of minTileSizes can be `ArrayRef<int64_t>`. We don't expect that the it would be modified. At the last part, you can do something like:

```
sizes.assign(minTileSizes);
std::replace_if(sizes.begin(), sizes...
```

File: compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp
Comment (hanhanW) [reply compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp]: (no action item required, just open the conversation for visibility.)

**Review by hanhanW — APPROVED** [review]
Thanks, just one nit! LGTM!

**Comment by c-rhodes** [comment]
> Thanks, just one nit! LGTM!

thanks for reviewing! Fixed the final nit 👍

File: compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp
Comment (c-rhodes) [reply compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp]: As Ben points out, doing this would require duplicating the logic which isn't ideal, particularly now that the precondition isn't specific to SME, so going to leave this as is.

**Pull request merged**
